python/py-xraylarch/Portfile

```diff
@@ -4,7 +4,7 @@ PortSystem          1.0
 PortGroup           python 1.0
 
 name                py-xraylarch
-version             0.9.58
+version             0.9.60
```

Comment:
why not the latest upstream version, which appears to be 0.9.61?

Reply:
I couldn't get it to build correctly, and I didn't understand why. Requirements?

Reply:
I couldn't get "it" = version 0.9.61 to build.

Reply:
what was the error?

Reply:
I think the following output (from `sudo port -vst install py310-xraylarch`) has the key error. Perhaps it will look familiar to you?

```
--->  Configuring py310-xraylarch
--->  Building py310-xraylarch
Executing:  cd "/opt/local/var/macports/build/_Users_fowlerj_Software_Portsnew_python_py-xraylarch/py310-xraylarch/work/xraylarch-0.9.61" && /opt/local/Library/Frameworks/Python.framework/Versions/3.10/bin/python3.10 setup.py --no-user-cfg build -j12 
/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/setuptools/installer.py:27: SetuptoolsDeprecationWarning: setuptools.installer is deprecated. Requirements should be satisfied by a PEP 517 installer.
  warnings.warn(
WARNING: The wheel package is not available.
/opt/local/Library/Frameworks/Python.framework/Versions/3.10/bin/python3.10: No module named pip.__main__; 'pip' is a package and cannot be directly executed
Traceback (most recent call last):
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/setuptools/installer.py", line 82, in fetch_build_egg
    subprocess.check_call(cmd)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/subprocess.py", line 369, in check_call
    raise CalledProcessError(retcode, cmd)
subprocess.CalledProcessError: Command '['/opt/local/Library/Frameworks/Python.framework/Versions/3.10/bin/python3.10', '-m', 'pip', '--disable-pip-version-check', 'wheel', '--no-deps', '-w', '/opt/local/var/macports/build/_Users_fowlerj_Software_Portsnew_python_py-xraylarch/py310-xraylarch/work/.tmp/tmpffb11ew2', '--quiet', '--index-url', "''", '--find-links', "''", 'setuptools_scm']' returned non-zero exit status 1.

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/local/var/macports/build/_Users_fowlerj_Software_Portsnew_python_py-xraylarch/py310-xraylarch/work/xraypy-xraylarch-2d3238f/setup.py", line 157, in <module>
    setup(name = 'xraylarch',
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/setuptools/__init__.py", line 86, in setup
    _install_setup_requires(attrs)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/setuptools/__init__.py", line 80, in _install_setup_requires
    dist.fetch_build_eggs(dist.setup_requires)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/setuptools/dist.py", line 876, in fetch_build_eggs
    resolved_dists = pkg_resources.working_set.resolve(
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/pkg_resources/__init__.py", line 789, in resolve
    dist = best[req.key] = env.best_match(
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/pkg_resources/__init__.py", line 1075, in best_match
    return self.obtain(req, installer)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/pkg_resources/__init__.py", line 1087, in obtain
    return installer(requirement)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/setuptools/dist.py", line 956, in fetch_build_egg
    return fetch_build_egg(self, req)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/setuptools/installer.py", line 84, in fetch_build_egg
    raise DistutilsError(str(e)) from e
distutils.errors.DistutilsError: Command '['/opt/local/Library/Frameworks/Python.framework/Versions/3.10/bin/python3.10', '-m', 'pip', '--disable-pip-version-check', 'wheel', '--no-deps', '-w', '/opt/local/var/macports/build/_Users_fowlerj_Software_Portsnew_python_py-xraylarch/py310-xraylarch/work/.tmp/tmpffb11ew2', '--quiet', '--index-url', "''", '--find-links', "''", 'setuptools_scm']' returned non-zero exit status 1.
Command failed:  cd "/opt/local/var/macports/build/_Users_fowlerj_Software_Portsnew_python_py-xraylarch/py310-xraylarch/work/xraylarch-0.9.61" && /opt/local/Library/Frameworks/Python.framework/Versions/3.10/bin/python3.10 setup.py --no-user-cfg build -j12 
Exit code: 1
```

Reply:
if it's trying to use `pip` then that typically indicates that you're missing a dependency. In this case it looks like that `toml` was added after 0.9.60 - see [upstream](https://github.com/xraypy/xraylarch/blame/0.9.61/requirements.txt)

Reply:
let's try and update to the latest version that was just released 0.9.62 and see if it is resolved. If not, I can take a look...

Reply:
Great idea...a new version is just what I was hoping for. At the moment, I cannot get it to download. We can give it some time for the PyPI to have the latest (which is now less than 1 hour old). I'll comment here once I know how 0.9.62 goes.

Reply:
I'm sorry. I cannot get 0.9.62 to be downloaded to even check the checksum. I just get 404 errors. Is there some special invocation I can use with `sudo port -d checksum py310-xraylarch` to help it find the necessary wheel file, or whatever? I get this error:

```
DEBUG: Executing org.macports.fetch (py310-xraylarch)
--->  xraylarch-0.9.62.tar.gz does not exist in /opt/local/var/macports/distfiles/py-xraylarch
--->  Attempting to fetch xraylarch-0.9.62.tar.gz from https://files.pythonhosted.org/packages/source/x/xraylarch
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0
DEBUG: Fetching distfile failed: The requested URL returned error: 404 
--->  Attempting to fetch xraylarch-0.9.62.tar.gz from https://ywg.ca.distfiles.macports.org/mirror/macports/distfiles/py-xraylarch
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0
DEBUG: Fetching distfile failed: The requested URL returned error: 404 Not Found
--->  Attempting to fetch xraylarch-0.9.62.tar.gz from http://ykf.ca.distfiles.macports.org/MacPorts/mpdistfiles/py-xraylarch
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
  0     0    0     0    0     0      0      0 --:--:--  0:00:01 --:--:--     0
DEBUG: Fetching distfile failed: The requested URL returned error: 404 Not Found
```
and so on.

Reply:
Still cannot get version 0.9.62 to be found by the usual methods. I tried adding a `PortGroup github` so the version 0.9.62 can be found. (Maybe that's just a temporary workaround.) That worked, but now I get a new error. Does this look at all familiar?

```
--->  Configuring py310-xraylarch
--->  Building py310-xraylarch
Executing:  cd "/opt/local/var/macports/build/_Users_fowlerj_Software_Portsnew_python_py-xraylarch/py310-xraylarch/work/xraylarch-0.9.62" && /opt/local/Library/Frameworks/Python.framework/Versions/3.10/bin/python3.10 setup.py --no-user-cfg build -j12 
Traceback (most recent call last):
  File "/opt/local/var/macports/build/_Users_fowlerj_Software_Portsnew_python_py-xraylarch/py310-xraylarch/work/xraypy-xraylarch-ccd03b6/setup.py", line 157, in <module>
    setup(name = 'xraylarch',
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/setuptools/__init__.py", line 87, in setup
    return distutils.core.setup(**attrs)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/setuptools/_distutils/core.py", line 109, in setup
    _setup_distribution = dist = klass(attrs)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/setuptools/dist.py", line 477, in __init__
    _Distribution.__init__(
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/setuptools/_distutils/dist.py", line 293, in __init__
    self.finalize_options()
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/setuptools/dist.py", line 901, in finalize_options
    ep(self)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/setuptools/dist.py", line 922, in _finalize_setup_keywords
    ep.load()(self, ep.name, value)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/setuptools_scm/integration.py", line 75, in version_keyword
    _assign_version(dist, config)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/setuptools_scm/integration.py", line 51, in _assign_version
    _version_missing(config)
  File "/opt/local/Library/Frameworks/Python.framework/Versions/3.10/lib/python3.10/site-packages/setuptools_scm/__init__.py", line 106, in _version_missing
    raise LookupError(
LookupError: setuptools-scm was unable to detect version for /opt/local/var/macports/build/_Users_fowlerj_Software_Portsnew_python_py-xraylarch/py310-xraylarch/work/xraypy-xraylarch-ccd03b6.

Make sure you're either building from a fully intact git repository or PyPI tarballs. Most other sources (such as GitHub's tarballs, a git checkout without the .git folder) don't contain the necessary metadata and will not work.

For example, if you're using pip, instead of https://github.com/user/proj/archive/master.zip use git+https://github.com/user/proj.git#egg=proj
Command failed:  cd "/opt/local/var/macports/build/_Users_fowlerj_Software_Portsnew_python_py-xraylarch/py310-xraylarch/work/xraylarch-0.9.62" && /opt/local/Library/Frameworks/Python.framework/Versions/3.10/bin/python3.10 setup.py --no-user-cfg build -j12 
Exit code: 1
```

Reply:
@reneeotten: [That was a good catch](https://github.com/xraypy/xraylarch/issues/380), realized that this problem was with lack of tarballs. I did not understand that from the upstream repository.

Is there any down-side to installing this by its wheel file? Would this require any change to the Portfile other than a build dependency on  py-wheel?

Reply:
one would need to override the `build.cmd` and `destroot.cmd` and such. There are examples on how to do that, I think `py-mkl` is one of them.

Reply:
Okay, thanks for the tip. I copied the Portfile from the simple `py-selenium`, with apparent success.

Reply:
I think this is ready. Can we merge?